src/services/services.ts

```diff
@@ -1801,6 +1801,9 @@ export function createLanguageService(
         };
         program = createProgram(options);
 
+        // unconditionally set oldProgram to undefined to prevent it from being captured in closure
+        options.oldProgram = undefined;

```

Comment:
@jakebailey @sheetalkamat What about this then? I admit the previous way to unreference oldProgram was confusing.

Reply:
This makes less sense... The thing I suggested (doing the same thing that the existing code did where you last edited) seems better.

Reply:
Ok I get it.

Reply:
Ok I played around with different approaches and could never figure out a clean way to fix this leak. I feel like there's indeed some funkiness in JSC that I just don't comprehend...

I will keep my fork for now since I cannot go forward without this ugly fix, but will try to comeback here if ever an update seems to fix anything...

I'm open to any other suggestion

Reply:
The most helpful thing would be a repro that we can run that shows this leak.

Reply:
JSC is technically a conservative collector, so that may tie into things. That said, it's also possible there's a leak for a precise GC without the mutation.

Reply:
I will build a little something to make the issue reproducible and figure out what could go on. But I will need some time. Please allow me 1-2 week and I will come back here with a clear workable playground.

Reply:
I've managed to setup a repro. Takes a while to get up and running, but hopefully you'll find a bit of time to check it out and maybe find an elegant fix 🙏

https://github.com/microsoft/TypeScript/issues/58137?notification_referrer_id=NT_kwDOAe993rQxMDE5MzE4ODQ5ODozMjQ3MjU0Mg#issuecomment-2076348479